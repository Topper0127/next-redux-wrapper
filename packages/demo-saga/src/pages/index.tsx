import React from 'react';
import {useSelector} from 'react-redux';
import {NextPage} from 'next';
import {State} from '../components/reducer';
import {SAGA_ACTION} from '../components/saga';

export interface ConnectedPageProps {
    custom: string;
}

// Page itself is not connected to Redux Store, it has to render Provider to allow child components to connect to Redux Store
const Page: NextPage<ConnectedPageProps> = ({custom}: ConnectedPageProps) => {
    const {page} = useSelector<State, State>(state => state);
    return (
        <div className="index">
            <pre>{JSON.stringify({page, custom}, null, 2)}</pre>
        </div>
    );
};

Page.getInitialProps = ({store}) => {
    store.dispatch({type: SAGA_ACTION});
    return {custom: 'custom'};
};

export default Page;
