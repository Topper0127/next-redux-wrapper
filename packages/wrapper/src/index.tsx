import React, {Component} from 'react';
import {Store} from 'redux';
import {NextComponentType, NextContext} from 'next';
import {NextAppContext} from 'next/app';

const _debug = false;
const DEFAULT_KEY = '__NEXT_REDUX_STORE__';
const isServer = typeof window === 'undefined';

const initStore = ({makeStore, initialState, config, ctx}: InitStoreOptions): Store => {
    const {storeKey} = config;

    const createStore = () =>
        makeStore(config.deserializeState(initialState), {
            ...ctx,
            ...config,
            isServer
        });

    if (isServer) return createStore();

    // Memoize store if client
    if (!window[storeKey]) {
        window[storeKey] = createStore();
    }

    return window[storeKey];
};

export default (makeStore: MakeStore, config?: Config) => {
    config = {
        storeKey: DEFAULT_KEY,
        debug: _debug,
        serializeState: state => state,
        deserializeState: state => state,
        ...config
    };

    return (App: NextComponentType | any) =>
        class WrappedApp extends Component<WrappedAppProps> {
            static displayName = `withRedux(${App.displayName || App.name || 'App'})`;

            static getInitialProps = async (appCtx: NextJSAppContext) => {
                if (!appCtx) throw new Error('No app context');
                if (!appCtx.ctx) throw new Error('No page context');

                const store = initStore({
                    makeStore,
                    config,
                    ctx: appCtx.ctx
                });

                if (config.debug)
                    console.log('1. WrappedApp.getInitialProps wrapper got the store with state', store.getState());

                appCtx.ctx.store = store;
                appCtx.ctx.isServer = isServer;

                let initialProps = {};

                if ('getInitialProps' in App) {
                    initialProps = await App.getInitialProps.call(App, appCtx);
                }

                if (config.debug) console.log('3. WrappedApp.getInitialProps has store state', store.getState());

                return {
                    isServer,
                    initialState: config.serializeState(store.getState()),
                    initialProps
                };
            };

            constructor(props, context) {
                super(props, context);

                const {initialState} = props;

                if (config.debug) console.log('4. WrappedApp.render created new store with initialState', initialState);

                this.store = initStore({
                    makeStore,
                    initialState,
                    config
                });
            }

            store: Store;

            render() {
                const {initialProps, initialState, ...props} = this.props;

                // Cmp render must return something like <Provider><Component/></Provider>
                return <App {...props} {...initialProps} store={this.store} />;
            }
        };
};

export interface Config {
    serializeState?: (any) => any;
    deserializeState?: (any) => any;
    storeKey?: string;
    debug?: boolean;
}

export interface NextJSContext extends NextContext {
    store: Store;
    isServer: boolean;
}

export interface NextJSAppContext extends NextAppContext {
    ctx: NextJSContext;
}

export interface MakeStoreOptions extends Config, NextJSContext {
    isServer: boolean;
}

export declare type MakeStore = (initialState: any, options: MakeStoreOptions) => Store;

export interface InitStoreOptions {
    makeStore: MakeStore;
    config: Config;
    initialState?: any;
    ctx?: NextJSContext;
}

export interface WrappedAppProps {
    initialProps: any; // stuff returned from getInitialProps
    initialState: any; // stuff in the Store state after getInitialProps
    isServer: boolean;
}

export interface AppProps {
    store: Store;
}
