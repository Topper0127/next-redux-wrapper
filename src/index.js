import React, {Component} from "react";

let _debug = false;
const DEFAULT_KEY = '__NEXT_REDUX_STORE__';
const isServer = typeof window === 'undefined';

/**
 * @param makeStore
 * @param initialState
 * @param config
 * @param ctx
 * @return {{getState: function, dispatch: function}}
 */
const initStore = ({makeStore, initialState, config, ctx = {}}) => {

    const {storeKey} = config;

    const createStore = () => makeStore(
        config.deserializeState(initialState),
        {
            ...ctx,
            ...config,
            isServer
        }
    );

    if (isServer) return createStore();

    // Memoize store if client
    if (!window[storeKey]) {
        window[storeKey] = createStore();
    }

    return window[storeKey];

};

/**
 * @param makeStore
 * @param config
 * @return {function(App): {getInitialProps, new(): WrappedApp, prototype: WrappedApp}}
 */
export default (makeStore, config = {}) => {

    config = {
        storeKey: DEFAULT_KEY,
        debug: _debug,
        serializeState: state => state,
        deserializeState: state => state,
        ...config
    };

    return (App) => (class WrappedApp extends Component {

        static displayName = `withRedux(${App.displayName || App.name || 'App'})`;

        static getInitialProps = async (appCtx) => {

            if (!appCtx) throw new Error('No app context');
            if (!appCtx.ctx) throw new Error('No page context');

            const store = initStore({
                makeStore,
                config,
                ctx: appCtx.ctx
            });

            if (config.debug) console.log('1. WrappedApp.getInitialProps wrapper got the store with state', store.getState());

            appCtx.ctx.store = store;
            appCtx.ctx.isServer = isServer;

            let initialProps = {};

            if ('getInitialProps' in App) {
                initialProps = await App.getInitialProps.call(App, appCtx);
            }

            if (config.debug) console.log('3. WrappedApp.getInitialProps has store state', store.getState());

            return {
                isServer,
                initialState: config.serializeState(store.getState()),
                initialProps: initialProps
            };

        };

        constructor(props, context) {

            super(props, context);

            const {initialState} = props;

            if (config.debug) console.log('4. WrappedApp.render created new store with initialState', initialState);

            this.store = initStore({
                makeStore,
                initialState,
                config
            });

        }

        render() {

            let {initialProps, initialState, ...props} = this.props;

            // Cmp render must return something like <Provider><Component/></Provider>
            return (
                <App {...props} {...initialProps} store={this.store}/>
            );

        }

    });

};