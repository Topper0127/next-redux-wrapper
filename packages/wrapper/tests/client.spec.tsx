/**
 * @jest-environment jsdom
 **/

import * as React from 'react';
import {create} from 'react-test-renderer';
import {makeStoreStub, DummyComponent, wrapper} from './testlib';
import {createWrapper} from '../src';
import {Store} from 'redux';
// import {act} from 'react-dom/test-utils';

const w: {testStoreKey: Store} = window as any;

describe('client integration', () => {
    afterEach(() => {
        delete w.testStoreKey;
    });

    describe('store taken from window', () => {
        beforeEach(() => {
            w.testStoreKey = makeStoreStub({state: 'val'})();
        });

        test('withRedux', async () => {
            const WrappedPage: any = wrapper.withRedux(DummyComponent);
            expect(create(<WrappedPage initialState={w.testStoreKey.getState()} />).toJSON()).toMatchSnapshot();
        });

        test('API functions', async () => {
            expect(await wrapper.getInitialPageProps()({} as any)).toMatchSnapshot();
        });
    });

    test('store is put to window when created', async () => {
        const wrapper = createWrapper(makeStoreStub({baz: 'qux'}), {storeKey: 'testStoreKey'});
        expect(await wrapper.getInitialPageProps()({} as any)).toMatchSnapshot();
        expect(w.testStoreKey.getState()).toMatchSnapshot();
    });
});
