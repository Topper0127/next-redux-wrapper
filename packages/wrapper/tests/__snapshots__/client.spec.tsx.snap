// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`client integration store is put to window when created 1`] = `
Object {
  "initialProps": Object {},
  "initialState": Object {
    "baz": "qux",
  },
}
`;

exports[`client integration store is put to window when created 2`] = `
Object {
  "baz": "qux",
}
`;

exports[`client integration store taken from window API functions 1`] = `
Object {
  "initialProps": Object {},
  "initialState": Object {
    "state": "val",
  },
}
`;

exports[`client integration store taken from window withRedux 1`] = `
<div>
  {"props":{},"state":{"state":"val"}}
</div>
`;
